### PR TITLE
network/ndisc: also drop configurations with infinite lifetime on RA …

### PR DESCRIPTION
**Pull request by yuwata**
…with zero lifetime being received

Follow-up for 4ccd48cb2693a3603f14a30b6286228597285d92.

Previously, even we received an RA with zero lifetime, configurations with infinite lifetime were not removed, as we call ndisc_drop_outdated() with USEC_INFINITY.

Now, when ndisc_drop_outdated() is called with USEC_INFINITY, all lifetime check are skipped, and all configurations provided by the matching router will be dropped.

### CODE REVIEW

**Comment by ssahani** [comment]
LGTM but we need to fix more too. The rfc says zero life time valid for only router not address and other resources. Hence dropping those part were wrong. We need to pull their indivisual lifetime and drop.

**Comment by yuwata** [comment]
Does it? Could you give me a reference?

**Comment by ssahani** [comment]
> Does it? Could you give me a reference?

Please see the commit message. Not able to copy via mobile phone.

**Comment by yuwata** [comment]
> > Does it? Could you give me a reference?
> 
> Please see the commit message. Not able to copy via mobile phone.

OK. I will check.

**Comment by ssahani** [comment]
> > > Does it? Could you give me a reference?
> > 
> > 
> > Please see the commit message. Not able to copy via mobile phone.
> 
> OK. I will check.
```
Router Lifetime
     16-bit unsigned integer.  The lifetime associated
     with the default router in units of seconds.  The
     field can contain values up to 65535 and receivers
     should handle any value, while the sending rules in
     Section 6 limit the lifetime to 9000 seconds.  A
     Lifetime of 0 indicates that the router is not a
     default router and SHOULD NOT appear on the default
     router list.  
-------------------------HERE-----------------------------
The Router Lifetime applies only to
     the router's usefulness as a default router; it
     does not apply to information contained in other
     message fields or options.  Options that need time
     limits for their information include their own
     lifetime fields.
---------------------TILL HERE------------------------
````

**Comment by keszybz** [comment]
I set `needs-rework` based on Susant's comments.

**Comment by yuwata** [comment]
> > > > Does it? Could you give me a reference?
> > > 
> > > 
> > > Please see the commit message. Not able to copy via mobile phone.
> > 
> > 
> > OK. I will check.
> 
> ```
> Router Lifetime
>      16-bit unsigned integer.  The lifetime associated
>      with the default router in units of seconds.  The
>      field can contain values up to 65535 and receivers
>      should handle any value, while the sending rules in
>      Section 6 limit the lifetime to 9000 seconds.  A
>      Lifetime of 0 indicates that the router is not a
>      default router and SHOULD NOT appear on the default
>      router list.  
> -------------------------HERE-----------------------------
> The Router Lifetime applies only to
>      the router's usefulness as a default router; it
>      does not apply to information contained in other
>      message fields or options.  Options that need time
>      limits for their information include their own
>      lifetime fields.
> ---------------------TILL HERE------------------------
> ```

Thanks, then your previous commit is RFC incompliant. I revert the commit. PTAL.

**Comment by ssahani** [comment]
> > > > > Does it? Could you give me a reference?
> > > > 
> > > > 
> > > > Please see the commit message. Not able to copy via mobile phone.
> > > 
> > > 
> > > OK. I will check.
> > 
> > 
> > ```
> > Router Lifetime
> >      16-bit unsigned integer.  The lifetime associated
> >      with the default router in units of seconds.  The
> >      field can contain values up to 65535 and receivers
> >      should handle any value, while the sending rules in
> >      Section 6 limit the lifetime to 9000 seconds.  A
> >      Lifetime of 0 indicates that the router is not a
> >      default router and SHOULD NOT appear on the default
> >      router list.  
> > -------------------------HERE-----------------------------
> > The Router Lifetime applies only to
> >      the router's usefulness as a default router; it
> >      does not apply to information contained in other
> >      message fields or options.  Options that need time
> >      limits for their information include their own
> >      lifetime fields.
> > ---------------------TILL HERE------------------------
> > ```
> 
> Thanks, then your previous commit is RFC incompliant. I revert the commit. PTAL.

Have not tested reverting  does not fix the original issue. It's same state as not dropping the zero lifetime. Hence we need  to process the zero lifetime precisely.  The original patch  https://github.com/systemd/systemd/commit/1b41aa9a82735960b0870b3b49f81deca5599526 fixes the issue.  https://github.com/systemd/systemd/issues/28426#issuecomment-1653398818

**Comment by yuwata** [comment]
As I commented in the original issue page, the issue should be in the hash and compare functions for `Route` object. I reopened the original issue page.

**Comment by ssahani** [comment]
> As I commented in the original issue page, the issue should be in the hash and compare functions for `Route` object. I reopened the original issue page.

Yes only for route . But we need to check for lifetime for each of the resources one by one like prefix, dns ... and drop if they also advertised to zero. If so we drop them. This is OK to go.

**Comment by yuwata** [comment]
> > As I commented in the original issue page, the issue should be in the hash and compare functions for `Route` object. I reopened the original issue page.
> 
> Yes only for route . But we need to check for lifetime for each of the resources one by one like prefix, dns ... and drop if they also advertised to zero. If so we drop them. This is OK to go.

All, except for captive portals are already checked and dropped if the lifetime is zero, IIRC.

**Comment by ssahani** [comment]
> > > As I commented in the original issue page, the issue should be in the hash and compare functions for `Route` object. I reopened the original issue page.
> > 
> > 
> > Yes only for route . But we need to check for lifetime for each of the resources one by one like prefix, dns ... and drop if they also advertised to zero. If so we drop them. This is OK to go.
> 
> All, except for captive portals are already checked and dropped if the lifetime is zero, IIRC.
Yeap I saw that . Thanks . I am starting work on https://github.com/systemd/systemd/issues/23674. I will drop you a mail for sync.

**Comment by yuwata** [comment]
> I am starting work on #23674. I will drop you a mail for sync.

mail? what? Sorry, I may overlook some discussion. I am often bad at managing my mail account. Feel free to include/exclude me.

**Comment by yuwata** [comment]
@keszybz @bluca PTAL.

File: src/network/networkd-ndisc.c [hunk 1]
```diff
@@ -885,6 +885,9 @@ static int ndisc_router_process_captive_portal(Link *link, sd_ndisc_router *rt)
         if (r < 0)
                 return log_link_warning_errno(link, r, "Failed to get router address from RA: %m");
 
+        /* RFC 4861 section 4.2. states that the lifetime in the message header should only used for the
```
Comment (bluca) [inline src/network/networkd-ndisc.c]: "be used"

File: src/network/networkd-ndisc.c [hunk 2]
```diff
@@ -885,6 +885,9 @@ static int ndisc_router_process_captive_portal(Link *link, sd_ndisc_router *rt)
...
+         * default gateway, but the captive portal option does not have lifetime field, hence, we use the
```
Comment (bluca) [inline src/network/networkd-ndisc.c]: "a lifetime"

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@bluca Thank you for the review. The comment is updated. Upgrading the green label.

**Pull request merged**